controllers/managedfusiondeployment_controller.go

```diff
@@ -0,0 +1,446 @@
+	agentCSVPrefix                       = "managed-fusion-agent"
+	monLabelKey                          = "app"
+	monLabelValue                        = "managed-fusion-agent"
+	pagerDutyKey                         = "pagerDutyServiceKey"
+	smtpPasswordKey                      = "smtpAuthPassword"
+)
+
+type ManagedFusionDeploymentReconciler struct {
+	Client             client.Client
+	UnrestrictedClient client.Client
+	Log                logr.Logger
+	Scheme             *runtime.Scheme
+	Namespace          string
+
+	ctx                           context.Context
+	prometheus                    *promv1.Prometheus
+	alertmanager                  *promv1.Alertmanager
+	alertmanagerConfig            *promv1a1.AlertmanagerConfig
+	alertmanagerConfigSecret      *corev1.Secret
+	managedFusionDeploymentSecret *corev1.Secret
+	smtpConfigData                *smtpConfig
+	pagerDutyConfigData           *pagerDutyConfig
+	CustomerNotificationHTMLPath  string
+}
+
+// Add necessary rbac permissions for managedfusiondeployment finalizer in order to set blockOwnerDeletion.
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources={alertmanagers,prometheuses,alertmanagerconfigs},verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources=prometheusrules,verbs=get;list;watch;create;update;patch
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources=podmonitors,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources=servicemonitors,verbs=get;list;watch;update;patch;create;delete
+// +kubebuilder:rbac:groups="",resources={secrets,secrets/finalizers},verbs=create;get;list;watch;update
+// +kubebuilder:rbac:groups=operators.coreos.com,resources=clusterserviceversions,verbs=get;list;watch;delete;update;patch
+// +kubebuilder:rbac:groups="apps",resources=statefulsets,verbs=get;list;watch
+// +kubebuilder:rbac:groups="coordination.k8s.io",resources=leases,verbs=create;get;list;watch;update
+
+// SetupWithManager creates an setup a ManagedFusionDeploymentReconciler to work with the provided manager
+func (r *ManagedFusionDeploymentReconciler) SetupWithManager(mgr ctrl.Manager, ctrlOptions *controller.Options) error {
+	if ctrlOptions == nil {
+		ctrlOptions = &controller.Options{
+			MaxConcurrentReconciles: 1,
+		}
+	}
+	// Filter the events for objects that are not owned by managed-fusion-agent-config secret
+	filterNonOwnedEvents := predicate.NewPredicateFuncs(
+		func(object client.Object) bool {
+			if object.GetName() == managedFusionAgentSecretName &&
+				object.GetNamespace() == r.Namespace {
+				return true
+			} else {
+				for _, owner := range object.GetOwnerReferences() {
+					if owner.Kind == "Secret" &&
+						owner.Name == managedFusionAgentSecretName {
+						return true
+					}
+				}
+			}
+			return false
+		},
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(*ctrlOptions).
+		For(&corev1.Secret{}).
```

Comment:
Even if I add a predicate for making sure we only acting on our managed fusion agent secret, it will not work because we are also owning secret object so the request will be queued if there are any changes in the secrets.

Reply:
It is ok to enqueue for secrets we own as long as we redirect into a config secret event. 
Why do you think this is a problem?

Reply:
Updated in latest commit